### PR TITLE
[varLib.models] Add interpolateFromValuesAndScalars()

### PR DESCRIPTION
**Pull request by behdad**
Faster than the alternative way to calculate them.

### CODE REVIEW

**Comment by behdad** [comment]
In theory I can use this in `varLib.merger`. Let me know if you want me to do that.

**Comment by F944** [comment]
Good

**Comment by justvanrossum** [comment]
Should `interpolateFromMasters()` then be implemented in terms of `interpolateFromMastersAndMasterScalars()`? That should be more efficient.

File: Lib/fontTools/varLib/models.py [hunk 1]
```diff
@@ -530,6 +531,22 @@ def interpolateFromMastersAndScalars(self, masterValues, scalars, *, round=noRou
         deltas = self.getDeltas(masterValues, round=round)
         return self.interpolateFromDeltasAndScalars(deltas, scalars)
 
+    @staticmethod
+    def interpolateFromMastersAndMasterScalars(masterValues, masterScalars):
```
Comment (justvanrossum) [inline Lib/fontTools/varLib/models.py:535]: This function is literally the same as `interpolateFromDeltasAndScalars()`, except for the variable names. Do we really need both?

**Comment by anthrotype** [comment]
I dont like the method name, it contains twice a name "master" which in theory we would like to avoid.
And if it is indeed faster (has anybody actually measured?), why do we need two methods to interpolate from sources and precomputed scalars? Ok backward compat sure, but if one of the two is definitely more efficient, then we should deprecate one and suggest to use the other.

**Comment by behdad** [comment]
All good points. Maybe I can just add a boolean optional argument to the existing API?

**Comment by behdad** [comment]
> I dont like the method name, it contains twice a name "master" which in theory we would like to avoid. And if it is indeed faster (has anybody actually measured?), why do we need two methods to interpolate from sources and precomputed scalars? Ok backward compat sure, but if one of the two is definitely more efficient, then we should deprecate one and suggest to use the other.

I think `interpolateFromMastersAndScalars` can be deprecated. I don't have a situation in mind where one has delta scalars but want to interpolate masters. I was going to merge the two with a bool parameter, but realized that one is a staticmethod and one isn't.

File: Lib/fontTools/varLib/models.py [hunk 2]
```diff
@@ -499,29 +500,34 @@ def getMasterScalars(self, targetLocation):
...
                 continue
-            contribution = delta * scalar
+            contribution = value * scalar
             if v is None:
                 v = contribution
             else:
                 v += contribution
         return v
 
+    @staticmethod
```
Comment (anthrotype) [inline Lib/fontTools/varLib/models.py:517]: maybe make it a @classmethod that takes `cls` as first param. it wouldn't change the way one one interacts with it vs staticmethod

File: Lib/fontTools/varLib/models.py [hunk 3]
```diff
@@ -530,6 +536,14 @@ def interpolateFromMastersAndScalars(self, masterValues, scalars, *, round=noRou
         deltas = self.getDeltas(masterValues, round=round)
         return self.interpolateFromDeltasAndScalars(deltas, scalars)
 
+    @staticmethod
+    def interpolateFromMastersAndMasterScalars(masterValues, masterScalars):
+        """Interpolate from master-values and master-scalars fetched
+        from getMasterScalars()."""
+        return VariationModel.interpolateFromValuesAndScalars(
```
Comment (anthrotype) [inline Lib/fontTools/varLib/models.py:543]: why can't simply add this `interpolateFromValuesAndScalars` staticmethod? What's the use of `interpolateFromMastersAndMasterScalars` -- besides the questionable name itself?
I think "value" is generic enough to mean "source" or "master" or even "delta", it's just the responsibility of the caller to pass in type of scalars that matches the kind of "values".

File: Lib/fontTools/varLib/models.py [hunk 4]
```diff
@@ -499,29 +500,34 @@ def getMasterScalars(self, targetLocation):
...
-            contribution = delta * scalar
+            contribution = value * scalar
             if v is None:
                 v = contribution
             else:
                 v += contribution
         return v
 
+    @staticmethod
+    def interpolateFromDeltasAndScalars(deltas, scalars):
```
Comment (anthrotype) [inline Lib/fontTools/varLib/models.py:518]: yes, this one we can keep as an alias of `interpolateFromValuesAndScalars` for backward compat, but I don't think we also need `interpolateFromMastersAndMasterScalars`

File: Lib/fontTools/varLib/models.py (hunk 2 above)
Comment (anthrotype) [reply Lib/fontTools/varLib/models.py:517]: or actually just get rid of it altogether (see my other comment)

File: Lib/fontTools/varLib/models.py (hunk 3 above)
Comment (behdad) [reply Lib/fontTools/varLib/models.py:543]: Sounds good to me. I'll adjust.

**Comment by behdad** [comment]
PTAL.

I didn't change staticmethod to classmethod. I don't understand the need for that.

**Comment by anthrotype** [comment]
> I don't understand the need for that.

I think the only difference/advantage is that the classmethod would work with subclasses, but it's ok if you prefer not to use

**Review by anthrotype — APPROVED** [review]

**Comment by F944** [comment]
H

**Pull request merged**